llvm/lib/Transforms/Instrumentation/MemProfiler.cpp

```diff
@@ -820,16 +820,26 @@ memprof::extractCallsFromIR(Module &M) {
           continue;
 
         StringRef CalleeName = CalledFunction->getName();
+        bool IsAlloc =
+            isAllocationWithHotColdVariant(CalledFunction, GetTLI(F));
         for (const DILocation *DIL = I.getDebugLoc(); DIL;
              DIL = DIL->getInlinedAt()) {
           StringRef CallerName = DIL->getSubprogramLinkageName();
           assert(!CallerName.empty() &&
                  "Be sure to enable -fdebug-info-for-profiling");
           uint64_t CallerGUID = IndexedMemProfRecord::getGUID(CallerName);
           uint64_t CalleeGUID = IndexedMemProfRecord::getGUID(CalleeName);
+          // Pretend that we are calling a function with GUID == 0 if we are
+          // calling a heap allocation function.
+          if (IsAlloc)
+            CalleeGUID = 0;
           LineLocation Loc = {GetOffset(DIL), DIL->getColumn()};
           Calls[CallerGUID].emplace_back(Loc, CalleeGUID);
           CalleeName = CallerName;
+          // FIXME: Recognize other frames that are associated with heap
+          // allocation functions.  It may be too early to reset IsAlloc to
+          // false here.
+          IsAlloc = false;
```

Comment:
Are you referring to the internal interceptor methods? These should be trimmed by the llvm-profdata logic here: https://github.com/llvm/llvm-project/blob/main/llvm/lib/ProfileData/MemProfReader.cpp#L209-L216

I think it's safe to assume the following:
* The operator new we want to replace will always be a direct call
* The internal frames of the memprof runtime will be dropped.

Do these assumptions change the logic you have here?

Reply:
Discussed offline.